scala/src/main/scala/com/xebia/functional/xef/scala/auto/package.scala

```diff
@@ -1,101 +1,97 @@
-    else serialName
-  List(CFunction(fnName, "Generated function for $fnName", Json.encodeJsonSchema(descriptor)))
-
-def addContext(docs: Iterable[String])(using scope: AIScope): Unit =
-  LoomAdapter.apply(scope.kt.addContext(docs.asJava, _))
+)(using
+    conversation: ScalaConversation
+): A =
+  val fromJson = new FromJson[A] {
+    def fromJson(json: String): A =
+      parse(json).flatMap(Decoder[A].decodeJson(_)).fold(throw _, identity)
+  }
+  conversation.prompt(chat, prompt, generateCFunctions.asJava, fromJson, promptConfiguration).join()
 
 def promptMessage(
-    prompt: String,
-    llmModel: Chat = OpenAI().DEFAULT_CHAT,
+    question: String,
+    chat: Chat = OpenAI.FromEnvironment.DEFAULT_CHAT,
     promptConfiguration: PromptConfiguration = PromptConfiguration.DEFAULTS
-)(using scope: AIScope): String =
-  LoomAdapter
-    .apply[String](
-      scope.kt.promptMessage(llmModel, prompt, promptConfiguration, _)
-    )
+)(using conversation: ScalaConversation): String =
+  conversation.promptMessage(chat, question, promptConfiguration).join()
 
 def promptMessages(
-    prompt: String,
-    llmModel: Chat = OpenAI().DEFAULT_CHAT,
-    functions: List[CFunction] = List.empty,
+    question: String,
+    chat: Chat = OpenAI.FromEnvironment.DEFAULT_CHAT,
+    functions: List[CFunction] = List(),
     promptConfiguration: PromptConfiguration = PromptConfiguration.DEFAULTS
-)(using scope: AIScope): List[String] =
-  LoomAdapter
-    .apply[java.util.List[String]](
-      scope.kt.promptMessages(llmModel, prompt, functions.asJava, promptConfiguration, _)
-    ).asScala.toList
-
-def pdf(
-    resource: String | File,
-    splitter: TextSplitter = TextSplitter.tokenTextSplitter(ModelType.getDEFAULT_SPLITTER_MODEL, 100, 50)
+)(using
+    conversation: ScalaConversation
 ): List[String] =
-  LoomAdapter
-    .apply[java.util.List[String]](count =>
-      resource match
-        case url: String => Loader.pdf(url, splitter.core, count)
-        case file: File => Loader.pdf(file, splitter.core, count)
-    ).asScala.toList
+  conversation.promptMessages(chat, question, functions.asJava, promptConfiguration).join().asScala.toList
+
+def promptStreaming(
+    question: String,
+    chat: Chat = OpenAI.FromEnvironment.DEFAULT_CHAT,
+    functions: List[CFunction],
+    promptConfiguration: PromptConfiguration
+)(using
+    conversation: ScalaConversation
+): LazyList[String] =
+  val publisher = conversation.promptStreaming(chat, question, promptConfiguration, functions.asJava)
+  val queue = new LinkedBlockingQueue[String]()
+  publisher.subscribe(new Subscriber[String] {
+    // TODO change to fs2 or similar
+    def onSubscribe(s: Subscription): Unit = s.request(Long.MaxValue)
+
+    def onNext(t: String): Unit = queue.add(t); ()
+
+    def onError(t: Throwable): Unit = throw t
+
+    def onComplete(): Unit = ()
+  })
+  LazyList.continually(queue.take)
```

Comment:
We should probably consider replacing this implementation or providing an alternative with something like fs2